### PR TITLE
Feat/governance

### PR DESCRIPTION
**Pull request by baroooo**
### Description

- Mento Governor contracts extending on OpenZeppelin's upgradeable governance implementation.
- It is compatible with Governor Bravo, similar to Rari implementation.
- Governor config will be updated once decided (#276)

Mainly consists of two components:
1. MentoGovernor.sol: Governing logic
2. TimelockController.sol: Actual owning entity that delays execution after a successful proposal

### Other changes

Make tests more isolated
ci checks PRs to feat/ branches

### Tested

Tested through unit tests

### Related issues

- Fixes https://github.com/mento-protocol/mento-general/issues/316

### Backwards compatibility

-

### Documentation

-

┆Issue is synchronized with this [Jira Story](https://mentolabs.atlassian.net/browse/MC-11) by [Unito](https://www.unito.io)

### CODE REVIEW

**Review by bayological — COMMENTED** [review]
This is not related to the changes but I noticed the CI checks are not being run. This is because the target branch is not develop and does not match the `feature/*` rule. Let's update the rules to include `feat/*`

File: test/governance/MentoGovernor/execute.t.sol [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    mentoGovernor.queue(targets, values, calldatas, keccak256(bytes(description)));
+
+    vm.roll(block.number + BLOCKS_DAY);
+    vm.warp(block.timestamp + 1 days);
+
+    assertEq(mockOwnable.protected(), 0);
+
+    _subject();
+
+    assertEq(mockOwnable.protected(), 1337);
```
Comment (bayological) [inline test/governance/MentoGovernor/execute.t.sol:86]: What is the significance of 1337 here? A comment would be useful to explain.

File: test/governance/MentoToken/burn.t.sol [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+pragma solidity 0.8.18;
+// solhint-disable func-name-mixedcase, contract-name-camelcase
+
+import { MentoToken_Test } from "./Base.t.sol";
+
+/**
+ * @notice Even though the burn function comes from OpenZeppelin's library,
+ * @notice this test assures correct integration.
+ */
+contract Burn_MentoToken_Test is MentoToken_Test {
```
Comment (bayological) [inline test/governance/MentoToken/burn.t.sol:11]: What do you think about adding more boundary tests? Specifically for calling burn from an address with no tokens and trying to burn zero tokens.

File: test/governance/MentoToken/burnFrom.t.sol [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    _newMentoToken();
+
+    deal(address(mentoToken), alice, initialBalance);
+  }
+
+  function _subject() internal {
+    mentoToken.burnFrom(alice, burnAmount);
+  }
+
+  function test_burnFrom_shouldRevert_whenNotAllowed() public {
```
Comment (bayological) [inline test/governance/MentoToken/burnFrom.t.sol:25]: How about adding a test case to verify that Alice cannot burn her/his/their own tokens without setting an allowance? Would that be useful?

File: test/governance/MentoToken/burnFrom.t.sol [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    vm.prank(bob);
+    vm.expectRevert("ERC20: insufficient allowance");
+    _subject();
+  }
+
+  function test_burnFrom_shouldBurnTokens_whenAllowed() public {
```
Comment (bayological) [inline test/governance/MentoToken/burnFrom.t.sol:31]: Is it worth also verifying the allowance post-burn is correct?

File: test/governance/MentoToken/burnFrom.t.sol [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+pragma solidity 0.8.18;
+// solhint-disable func-name-mixedcase, contract-name-camelcase
+
+import { MentoToken_Test } from "./Base.t.sol";
+
+/**
+ * @notice Even though the burn function comes from OpenZeppelin's library,
+ * @notice this test assures correct integration.
+ */
+contract BurnFrom_MentoToken_Test is MentoToken_Test {
```
Comment (bayological) [inline test/governance/MentoToken/burnFrom.t.sol:11]: How about adding a case to verify trying to `burnFrom` more tokens than Alice's balance but within the approved allowance reverts

**Review by bayological — APPROVED** [review]
LGTM 🚣 ⛰️. Some minor comments on test naming and potential test cases that could be added.

File: contracts/governance/MentoGovernor.sol [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * @title Mento Governor
+ * @author Mento Labs
+ * @notice Governor contract extending on OpenZeppelin's upgradeable governance contracts.
+ */
+contract MentoGovernor is
+  GovernorUpgradeable,
+  GovernorSettingsUpgradeable,
+  GovernorCompatibilityBravoUpgradeable,
```
Comment (bowd) [inline contracts/governance/MentoGovernor.sol:22]: Is there a requirement from the LockingContract side to be compatible to Governor Bravo?

File: contracts/governance/TimelockController.sol [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+   * @param executors List of addresses that are allowed to execute proposals. address(0) can be used to allow any account.
+   * @param admin The admin address that will be used to set the proposer role and then will be renounced.
+   */
+  function __MentoTimelockController_init(
+    uint256 minDelay,
+    address[] memory proposers,
+    address[] memory executors,
+    address admin
+  ) external initializer {
+    __TimelockController_init(minDelay, proposers, executors, admin);
```
Comment (bowd) [inline contracts/governance/TimelockController.sol:28]: Should we directly add a `canceller` here to make it simple?
Basically call the `_setupRole(CANCELLER_ROLE, canceller)` and add this `canceller` as a parameter which would be the Mento community multisig.

**Review by bowd — COMMENTED** [review]
Left a few comments now but I'll review the tests after you revert the changes as we talked on Friday

File: test/governance/Emission/emitTokens.t.sol [hunk 8]
```diff
@@ -1,187 +1,135 @@
...
-    emission.renounceOwnership();
-
-    assertEq(emission.owner(), address(0));
+  function _subject() internal returns (uint256) {
+    return emission.emitTokens();
   }
 
   function test_emitTokens_whenAfter1Month_shouldMintCorrectAmountToTarget() public {
-    _setupEmissionContract();
     uint256 calculatedAmountFor1Month = 3_692_586_569039559708901376;
```
[the remark this answers is not in the document]
Comment (baroooo) [reply test/governance/Emission/emitTokens.t.sol:21]: Using the python script: https://www.notion.so/Spike-Emission-contract-a3389802add2438ca482c6d18a607a2f

File: test/governance/MentoGovernor/execute.t.sol (hunk 1 above)
Comment (baroooo) [reply test/governance/MentoGovernor/execute.t.sol:86]: Just a random number that updates the protected state var.

File: test/governance/MentoToken/burn.t.sol (hunk 2 above)
Comment (baroooo) [reply test/governance/MentoToken/burn.t.sol:11]: Not sure if token related boundary tests would be beneficial as we use OZ implementation without changes.

File: contracts/governance/MentoGovernor.sol (hunk 6 above)
Comment (baroooo) [reply contracts/governance/MentoGovernor.sol:22]: There is no requirements, I guess the main purpose they used it was to be compatible with the old tools that supports Bravo. Should we replace it with a counting module ?

**Comment by baroooo** [comment]
Thanks @bayological and @bowd . I was planning to revert test related changes in the next PR to avoid confusion. I will commit them here instead so we can be done with governance contracts.

File: test/governance/MentoToken/burnFrom.t.sol (hunk 5 above)
Comment (baroooo) [reply test/governance/MentoToken/burnFrom.t.sol:11]: Similar to the comment above, since we use the OZ implementation not sure if we should test token related functionality.
Current burn and mint related tests are in place to verify if we integrated burnable extension and mint functionality(ie, changes we introduced) correctly.

File: test/governance/MentoGovernor.t.sol [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+    (uint256 proposalId, , , , ) = _proposeCallProtectedFunction();
+
+    vm.roll(block.number + votingDelay + 1);
+
+    vm.prank(bob);
+    mentoGovernor.castVote(proposalId, 1);
+    vm.prank(charlie);
+    mentoGovernor.castVote(proposalId, 0);
+
+    vm.roll(block.number + votingPeriod);
```
Comment (bowd) [inline test/governance/MentoGovernor.t.sol:150]: Very tiny nit: `vm.skip(votingPeriod)` is the same as `vm.roll(block.number + votingPeriod)`

**Review by bowd — APPROVED** [review]

File: test/governance/MentoGovernor.t.sol (hunk 9 above)
Comment (baroooo) [reply test/governance/MentoGovernor.t.sol:150]: I think skip is for timestamp and not the block. I replaced timestamps instead.
